src/js/links.js

```diff
@@ -0,0 +1,251 @@
+      }
+      return minLinks ? minLinks.shift() : null;
+    }
+
+    var finalLinks = [];
+    for (var nextLink = getNextLink();
+         nextLink && finalLinks.length < this.maxNumLinks;
+         nextLink = getNextLink()) {
+      finalLinks.push(nextLink);
+    }
+
+    return finalLinks;
+  },
+
+  /**
+   * Gets the current set of links contained in the grid.
+   *
+   * @return {Object} The links in the grid.
+   */
+  getLinks() {
+    var pinnedLinks = Array.slice(gPinnedLinks.links);
+    var links = this._getMergedProviderLinks();
+
+    var sites = new Set();
+    for (var link of pinnedLinks) {
+      if (link) {
+        sites.add(ProviderManager.extractSite(link.url));
+      }
+    }
+
+    // Filter blocked and pinned links and duplicate base domains.
+    links = links.filter(function(link) {
+      var site = ProviderManager.extractSite(link.url);
+      if (site === null || sites.has(site)) {
+        return false;
+      }
+      sites.add(site);
+
+      return !gBlockedLinks.isBlocked(link) && !gPinnedLinks.isPinned(link);
+    });
+
+    // Try to fill the gaps between pinned links.
+    for (var i = 0; i < pinnedLinks.length && links.length; i++) {
+      if (!pinnedLinks[i]) {
+        pinnedLinks[i] = links.shift();
+      }
+    }
+
+    // Append the remaining links if any.
+    if (links.length) {
+      pinnedLinks = pinnedLinks.concat(links);
+    }
+
+    for (link of pinnedLinks) {
+      if (link) {
+        link.baseDomain = ProviderManager.extractSite(link.url);
+      }
+    }
+
+    // Get the set of enhanced links (if any) from the Directory Links Provider.
+    var enhancedLinks = [];
+    for (link of pinnedLinks) {
+      if (link) {
+        enhancedLinks.push(DirectoryLinksProvider.getEnhancedLink(link));
+      }
+    }
+
+    return {links: pinnedLinks, enhancedLinks};
+  },
+
+  _incrementSiteMap: function(map, link) {
+    if (gBlockedLinks.isBlocked(link)) {
+      // Don't count blocked URLs.
+      return;
+    }
+    var site = ProviderManager.extractSite(link.url);
+    map.set(site, (map.get(site) || 0) + 1);
+  },
+
+  /**
+   * Calls getLinks on the given provider and populates our cache for it.
+   *
+   * @param {Provider} aProvider The provider whose cache will be populated.
+   * @param {Object} cache The provider cache to be populated.
+   * @param {Boolean} aForce When true, populates the provider's cache even when it's
+   *               already filled.
+   */
+  _populateProviderCache: async(function*(aProvider, cache, aForce) {
+    var createCache = !cache;
+    if (createCache) {
+      cache = {
+        // Start with a resolved promise.
+        populatePromise: new Promise(resolve => resolve()),
+      };
+      Links._providers.set(aProvider, cache);
+    }
+    // Chain the populatePromise so that calls are effectively queued.
+    cache.populatePromise = cache.populatePromise.then(() => {
+      return new Promise(resolve => {
+        if (!createCache && !aForce) {
```

Comment:
Ok, maybe I missed something, but I couldn't find any reason for this code to run asycn? If that is true, then it could just become:

```JS
_populateProviderCache: function(aProvider, cache, aForce) {
  if (!cache) {
    cache = {};
    Links._providers.set(aProvider, cache);
  }
  var links = aProvider.getLinks()
    // Filter out null and undefined links so we don't have to deal with
    // them in getLinks when merging links from providers.
    .filter(link => link)
  cache.sortedLinks = links;
  cache.siteMap = links.reduce((map, link) => {
    Links._incrementSiteMap(map, link);
    return map;
  }, new Map());
  cache.linkMap = links.reduce((map, link) => {
    map.set(link.url, link);
    return map;
  }, new Map());
},

```